press/press/doctype/deploy_candidate/deploy_candidate.py

```diff
@@ -778,6 +780,78 @@ def get_dependency_version(self, dependency):
 		version = find(self.dependencies, lambda x: x.dependency == dependency).version
 		return f"{dependency} {version}"
 
+	def get_pull_update_dict(self) -> dict[str, PullUpdate]:
+		"""
+		Returns a dict of apps with:
+
+		`old` hash: for which there already exist cached layers from previously
+		deployed Benches that have been created from this Deploy Candidate.
+
+		`new` hash: which can just be 'git pull' updated, i.e. a new layer does
+		not need to be built for them from scratch.
+		"""
+
+		# Deployed Benches from current DC with (potentially) cached layers
+		benches = frappe.get_list(
+			"Bench", filters={"group": self.group, "status": "Active"}, limit=1
+		)
+		if not benches:
+			return {}
+
+		bench_name = benches[0]["name"]
+
+		# db.sql used cause get_list wasn't working as expected
+		bench_apps = frappe.db.sql(
+			"SELECT `app`, `source`, `release`, `hash` FROM `tabBench App` WHERE parent=%s",
+			(bench_name),
+			as_dict=1,
+		)
+
+		update_app_hashes = {app.app: app.hash for app in self.apps}
+		pull_update: dict[str, PullUpdate] = {}
+
+		for app in bench_apps:
+			release: AppRelease = frappe.get_doc("App Release", app["release"])
+			app_name = app["app"]
+			update_hash = update_app_hashes[app_name]
+
+			if release.hash == update_hash:
+				continue
+
+			file_diff = release.get_changed_files_against_hash(update_hash)
+			if not can_pull_update(file_diff):
+				"""
+				If current app is not being pull_updated, then no need to
+				pull update apps later in the sequence.
+
+				This is because once an image layer hash changes all layers
+				after it have to be rebuilt.
+				"""
+				break
+
+			pull_update[app_name] = {"old": release.hash, "new": update_hash}
+		return pull_update
+
+
+def can_pull_update(changed_files: list[str]) -> bool:
+	for file in changed_files:
+		# Requires database migration
+		if file.endswith(".json") and "doctype" in file:
+			return False
```

Comment:
Something that might require database migration doesn't really require image rebuild?

Reply:
Right, I conflated new site with bench init.